modules/01_Theory.md

```diff
@@ -0,0 +1,409 @@
+
+```html
+<html>
+  <head>
+    <script src="a.js"></script>
+    <script src="b.js"></script>
+  </head>
+  <body></body>
+</html>
+```
+
+This approach had some downsides, especially as web pages grew larger and more complex. In particular, all scripts loaded onto the same page share the same scope—appropriately called the “global scope”—meaning the scripts had to be very careful not to overwrite each others’ variables and functions.
+
+Any system that solves this problem by giving files their own scope while still providing a way to make bits of code available to other files can be called a “module system.” (It may sound obvious to say that each file in a module system is called a “module,” but the term is often used to contrast with _script_ files, which run outside a module system, in a global scope.)
+
+> There are [many module systems](https://github.com/myshov/history-of-javascript/tree/master/4_evolution_of_js_modularity), and TypeScript [supports emitting several](https://www.typescriptlang.org/tsconfig/#module), but this documentation will focus on the two most important systems today: ECMAScript modules (ESM) and CommonJS (CJS).
+>
+> ECMAScript Modules (ESM) is the module system built into the language, supported in modern browsers and in Node.js since v12. It uses dedicated `import` and `export` syntax:
+>
+> ```js
+> // a.js
+> export default "Hello from a.js";
+> ```
+>
+> ```js
+> // b.js
+> import a from "./a.js";
+> console.log(a); // 'Hello from a.js'
+> ```
+>
+> CommonJS (CJS) is the module system that originally shipped in Node.js, before ESM was part of the language specification. It’s still supported in Node.js alongside ESM. It uses plain JavaScript objects and functions named `exports` and `require`:
+>
+> ```js
+> // a.js
+> exports.message = "Hello from a.js";
+> ```
+>
+> ```js
+> // b.js
+> const a = require("./a");
+> console.log(a.message); // 'Hello from a.js'
+> ```
+
+Accordingly, when TypeScript detects that a file is a CommonJS or ECMAScript module, it starts by assuming that file will have its own scope. Beyond that, though, the compiler’s job gets a little more complicated.
+
+## TypeScript’s job concerning modules
+
+The TypeScript compiler’s chief goal is to look at input code and tell the author about problems the output code might encounter at runtime. To do this, the compiler needs to know some things about the code’s intended runtime environment—what globals are available, for example. When the input code is a module, there are several additional questions the compiler needs to answer in order to do its job. Let’s use a few lines of input code as an example to think about all the information needed to analyze it:
+
+```ts
+import sayHello from "greetings";
+sayHello("world");
+```
+
```

Comment:
This section is good but sort of puts the cart before the horse. We can start with "At runtime, something will happen" (via all these same steps described in checking), then describe TS's job in terms of asking questions about what that behavior is.

Reply:
Hm, that’s a good point, but this is supposed to be `.ts` file contents, so it’s hard to say what’s going to happen at runtime without presupposing what module format we’re going to emit. But maybe it can be stated abstractly enough to gloss over the potential ESM/CJS differences here.

Reply:
The [updated version](https://delightful-glacier-09aa73710-2946.centralus.azurestaticapps.net/docs/handbook/modules/theory.html#typescripts-job-concerning-modules) of this section was rewritten to account for the possibility that the runtime may load the TS file directly, which further complicates prefacing this with a discussion of what the runtime will do 🤔